### ISSUE TITLE
Prompt buffer needs 3 `<Esc>` presses to leave Insert mode after callback temporarily leaves window

### ISSUE THREAD

**Issue by seandewar** [vim/vim#9212]
### Steps to reproduce

Run the following:

```vim
new
set showmode buftype=prompt
startinsert
call timer_start(0, {-> execute("wincmd p|wincmd p", "")})
```

Afterwards, trying to return to Normal mode with `<Esc>` will instead act like `<C-O>`, showing `-- (insert) --` as the mode.
Pressing `<Esc>` another time will put you back in Insert mode, requiring a third `<Esc>` to get to the usual Normal mode.

Looks like the below logic in `leaving_window` is responsible, as removing it solves the issue:
```c
    // When leaving the window (or closing the window) was done from a
    // callback we need to break out of the Insert mode loop and restart Insert
    // mode when entering the window again.
    if (State & INSERT)
    {
	stop_insert_mode = TRUE;
	if (win->w_buffer->b_prompt_insert == NUL)
	    win->w_buffer->b_prompt_insert = 'A';
    }
```

Removing this block didn't cause any issues that I could see, but I didn't do a lot of testing (also the tests in `test_prompt_buffer.vim` still pass at least, but maybe its lacking a relevant regression test there).

### Expected behaviour

Pressing `<Esc>` once should return to Normal mode as usual.

### Operating system

Fedora 35, GNOME 41.1

### Version of Vim

Vim v8.2.3668, GNOME Terminal 3.42.1

**Comment by brammool** [comment]
The problem is that when coming back to the prompt window we are still in Insert mode, since we didn't get to the main loop in between the commands. The solution is to not set restart_edit when already in Insert mode.

**Comment by chrisbra** [comment]
I suppose, fixed by 34c20ff85b87be587ea5d0398812441b502ee6a5
